compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp

```diff
@@ -21,42 +21,120 @@
+      map = map.compose(llvm::cast<AffineMapAttr>(mapsAttr[i]).getAffineMap());
+    }
+    return map;
+  }
+  return AffineMap();
+}
+
 EncodingAttr EncodingAttr::get(MLIRContext *ctx, int64_t operandIndex,
                                EncodingOpType opType, ArrayRef<Type> elemTypes,
                                ArrayRef<AffineMap> maps,
-                               std::optional<AffineMap> bcastMap,
                                ArrayRef<int64_t> roundDimsTo,
                                ArrayRef<Attribute> layouts) {
   Builder b(ctx);
   auto opTypeAttr = EncodingOpTypeAttr::get(ctx, opType);
   auto roundDimsToAttr = roundDimsTo.empty()
                              ? DenseI64ArrayAttr()
                              : b.getDenseI64ArrayAttr(roundDimsTo);
-  auto bcastMapAttr = bcastMap.has_value()
-                          ? AffineMapAttr::get(bcastMap.value())
-                          : AffineMapAttr();
   auto layoutsAttr = layouts.empty() ? ArrayAttr() : b.getArrayAttr(layouts);
   return get(ctx, b.getIndexAttr(operandIndex), opTypeAttr,
              b.getTypeArrayAttr(elemTypes), b.getAffineMapArrayAttr(maps),
-             bcastMapAttr, roundDimsToAttr, layoutsAttr);
+             roundDimsToAttr, layoutsAttr);
 }
 
-AffineMap EncodingAttr::getMapForOperandIndex() const {
-  auto index = getOperandIndex().getValue().getZExtValue();
-  switch (index) {
-  case MATMUL_LHS:
-  case MATMUL_RHS:
-  case MATMUL_RESULT: {
-    auto indexingMap =
-        llvm::cast<AffineMapAttr>(getUserIndexingMaps()[index]).getAffineMap();
-    if (auto bcastMap = getBcastMap()) {
-      indexingMap = bcastMap.getAffineMap().compose(indexingMap);
+static bool isNestedAffineMapOrArrayAttr(Attribute attr, unsigned maxDepth) {
+  if (isa<AffineMapAttr>(attr)) {
+    return true;
+  }
+  if (auto arrayAttr = llvm::dyn_cast<ArrayAttr>(attr)) {
+    if (maxDepth == 0) {
+      return false;
+    }
+    return llvm::all_of(arrayAttr.getValue(), [&](Attribute nestedAttr) {
+      return isNestedAffineMapOrArrayAttr(nestedAttr, maxDepth - 1);
+    });
+  }
+  return false;
+}
+
+LogicalResult
+EncodingAttr::verify(function_ref<mlir::InFlightDiagnostic()> emitError,
+                     IntegerAttr operandIndexAttr,
+                     EncodingOpTypeAttr opTypeAttr, ArrayAttr elementTypesAttr,
+                     ArrayAttr userIndexingMapsAttr,
+                     DenseArrayAttr roundDimsToAttr, ArrayAttr layoutsAttr) {
+  if (userIndexingMapsAttr) {
+    unsigned index = operandIndexAttr.getValue().getZExtValue();
+    if (index >= userIndexingMapsAttr.size()) {
+      return emitError()
+             << "`operandIndex` exceeds the size of `user_indexing_maps`";
+    }
+    if (!isNestedAffineMapOrArrayAttr(userIndexingMapsAttr, 2)) {
+      return emitError() << "`user_indexing_maps` should be a nested array of "
+                            "`AffineMapAttr` with max 2 nesting levels";
```

Comment:
I'd drop the maxDepth limit check and verify if the maps are composable or not. It looks more robust to me and we can have it documented in `EncodingAttrs.td`. What do you think?

Reply:
Yeah, I changed it to use `getComposedAffineMap` instead.